include/eosio/vm/wasm_stack.hpp

```diff
@@ -7,45 +7,65 @@
 
 namespace eosio { namespace vm {
-   template <size_t Elems, typename ElemT, typename Allocator>
-   class fixed_stack {
+   template <typename ElemT, size_t ElemSz, typename Allocator = nullptr_t >
+   class stack {
     public:
-      fixed_stack(Allocator& alloc) : _s(managed_vector<ElemT, Allocator>{ alloc, Elems }) {}
-      void        push(ElemT e) { _s[_index++] = e; }
+      template <typename Alloc=Allocator, typename = std::enable_if_t<std::is_same_v<Alloc, nullptr_t>, int>>
+      stack() 
+         : _store(unmanaged_vector<ElemT>{ ElemSz }) {}
+
+      template <typename Alloc=Allocator, typename = std::enable_if_t<!std::is_same_v<Alloc, nullptr_t>, int>>
+      stack(Alloc& alloc) 
+         : _store(managed_vector<ElemT, Alloc>{alloc, ElemSz }) {}
+
+      void push(ElemT e) { 
+         _store[_index++] = e; 
+      }
+
       ElemT& get(uint32_t index) const {
          EOS_VM_ASSERT(index <= _index, wasm_interpreter_exception, "invalid stack index");
-         return _s[index];
+         return (ElemT&)_store[index];
       }
       void set(uint32_t index, const ElemT& el) {
          EOS_VM_ASSERT(index <= _index, wasm_interpreter_exception, "invalid stack index");
-         _s[index] = el;
+         _store[index] = el;
       }
-      ElemT        pop() { return _s[--_index]; }
-      void         eat(uint32_t index) { _index = index; }
+      ElemT pop() { return _store[--_index]; }
+      void  eat(uint32_t index) { _index = index; }
       // compact the last element to the element pointed to by index
-      void         compact(uint32_t index) { 
-         _s[index] = _s[_index-1];
+      void compact(uint32_t index) { 
+         _store[index] = _store[_index-1];
          _index = index+1;
       }
       uint16_t     current_index() const { return _index; }
-      ElemT&       peek() { return _s[_index - 1]; }
-      const ElemT& peek() const { return _s[_index - 1]; }
-      ElemT&       peek(size_t i) { return _s[_index - 1 - i]; }
-      ElemT        get_back(size_t i) { return _s[_index - 1 - i]; }
+      ElemT&       peek() { return _store[_index - 1]; }
+      const ElemT& peek() const { return _store[_index - 1]; }
+      ElemT&       peek(size_t i) { return _store[_index - 1 - i]; }
+      ElemT        get_back(size_t i) { return _store[_index - 1 - i]; }
       void         trim(size_t amt) { _index -= amt; }
       uint16_t     size() const { return _index; }
 
     private:
-      managed_vector<ElemT, Allocator> _s;
-      uint16_t                         _index = 0;
+      template <typename Elem, typename Alloc>
+      struct base_data_store {
+         using type = managed_vector<Elem, Alloc>;
+      };
+      template <typename Elem>
+      struct base_data_store <Elem, nullptr_t> {
+         using type = unmanaged_vector<Elem>;
+      };
+
+      using base_data_store_t = typename base_data_store<ElemT, Allocator>::type;
+
+      base_data_store_t _store;
+      uint16_t          _index = 0;
```

Comment:
This imposes a limit of 64536 elements, which will silently wrap instead of causing an error.